### PR TITLE
feat(gear-programs): vft-treasury program implementation

### PR DESCRIPTION
**Pull request by playX18**

### CODE REVIEW

File: gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs:9]:
```suggestion
#[test]
fn test_lock_unlock_assets() {
```

File: gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs:39]:
```suggestion
#[test]
fn test_lock_unlock_assets_two_accounts() {
```

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/mod.rs:87]:
```suggestion
        let source = self.exec_context.actor_id();
```

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/mod.rs:90]: Call to a `vft` program could fail, sou you should process this case (wait for a reply)

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/mod.rs:104]:
```suggestion
    pub async fn withdraw(&mut self, token: H160, recepient: ActorId, amount: U256) {
```
This method is called by `ethereum-event-client` which knows nothing about `vara` token id(because ethereum smart-contract works only with `erc20` tokens), so you should store mapping `ethereum token address(=H160) <-> vara token address(=ActorId)` and fetch `vara` token id based on it

Comment (mertwole) [reply gear-programs/vft-treasury/src/services/mod.rs:104]: Also admin (which will be a governance in future) should be able to add/remove tokens from this mapping

**Review by mertwole — CHANGES_REQUESTED** [review]

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (ByteNacked) [inline gear-programs/vft-treasury/src/services/mod.rs:86]: I think we should pass the sender's address here:


    pub async fn deposit(&mut self, sender: ActorId, token: ActorId, amount: U256, to: H160) {

because I won't be transferring tokens to the bridge-payment address.

Comment (ByteNacked) [reply gear-programs/vft-treasury/src/services/mod.rs:86]: cc @mertwole

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (ByteNacked) [reply gear-programs/vft-treasury/src/services/mod.rs:90]: Should we also return `Result` in `deposit` method for better clarity (same as in `vtf-gateway`)?
 
 ```rust
pub async fn deposit(&mut self, token: ActorId, amount: U256, to: H160) -> Result<Error> 
```

For ex. return errors `NotEnoughFunds`, `BadParams`, etc.

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (mertwole) [reply gear-programs/vft-treasury/src/services/mod.rs:86]: yes, agree with that

Comment (ByteNacked) [reply gear-programs/vft-treasury/src/services/mod.rs:86]: Also `deposit` method should probably return mapped `eth_token_id`, 
it will be used in `bridging-payment` `BridgingPaymentEvents` event.

Comment (mertwole) [inline gear-programs/vft-treasury/src/services/mod.rs:86]: The logic of sending message to bridge built-in actor is missing

Comment (mertwole) [reply gear-programs/vft-treasury/src/services/mod.rs:86]: Also `nonce` will be used by `relayer` to distinguish sent messages

**Comment by playX18** [comment]
Copied msg tracker and reply handling logic from vft-gateway and refactored treasury to send messages to builtin bridge + added error handling

File: gear-programs/vft-treasury/src/wasm/Cargo.toml
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/Cargo.toml:24]:
```suggestion
blake2.workspace = true
```

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/mod.rs:248]: Anyone can submit request, not only `bridging-payment`

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/mod.rs:28]: I think it'd be better to store just `Vec<H160, ActorId>` and perform mapping using linear search. It's not significantly slower and will allow not to maintain 2 mappings

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/mod.rs:157]: Maybe some method `ensure_admin`?

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/mod.rs:214]: Why do you need it?

File: gear-programs/vft-treasury/src/services/msg_tracker.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/msg_tracker.rs:18]:
```suggestion
    DepositToTreasury {
```
There can be other tokens besides wVARA

File: gear-programs/vft-treasury/src/services/msg_tracker.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/msg_tracker.rs:45]: Isn't it a error when we cannot find corresponding message?

File: gear-programs/vft-treasury/src/services/token_operations.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/token_operations.rs:5]: Don't leave commented out code

File: gear-programs/vft-treasury/src/services/utils.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/utils.rs:5]: It't not very clear from this fn signature what should it do

File: gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs:29]: Please move it after `Program::vft_treasury(&system);` as it's a setup logic

File: gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs:44]: Assert here that balance have changed

File: gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs:8]: I'd add some more tests, for example:
- check that when there's no token present in mapping it'll return error and not transfer tokens to treasury
- check that `withdraw` can be called only be `ethereum-event-cient`
- check that anyone can call `deposit` after token owner have approved transfer (as it's our use-case with `bridging-payment`)

**Review by mertwole — CHANGES_REQUESTED** [review]

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (playX18) [reply gear-programs/vft-treasury/src/services/mod.rs:164]: @mertwole how would it be different? Should `admin` be `Option<ActorId>` in that case?

Comment (mertwole) [reply gear-programs/vft-treasury/src/services/mod.rs:164]: I think the logic will be a little more clear

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/mod.rs:178]: Please also check there that there're no vara address that maps to multiple eth addresses and no eth address mapped to multiple vara ones

File: gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs:62]: Can initialization be re-used across tests?

File: gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs:97]: maybe `test_withdraw_fails_with_bad_origin`?

Comment (mertwole) [reply gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs:97]: And as I see you don't assert that it's failed

File: gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs:125]: It should work when `account1_id` don't have tokens on his balance

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (ByteNacked) [inline gear-programs/vft-treasury/src/services/mod.rs:402]: Looks like a type, should be ` data.gear_bridge_builtin,` address

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (ByteNacked) [inline gear-programs/vft-treasury/src/services/mod.rs:255]: It seems this contract should not depend on `bridging_payment_service` address so it should be removed

File: gear-programs/vft-treasury/src/services/error.rs
Comment (ByteNacked) [inline gear-programs/vft-treasury/src/services/error.rs:32]: Some errors are not used in the contract. Please either remove or use the unused ones.

File: gear-programs/vft-treasury/src/services/utils.rs
Comment (ByteNacked) [inline gear-programs/vft-treasury/src/services/utils.rs:95]: Do we really need a `BridgeBuiltinStep` status, it seems duplicate of `MessageStatus::TokenTransferCompleted(true)`?

**Review by ByteNacked — CHANGES_REQUESTED** [review]
Good job overall, please check some suggestions in comments.

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/mod.rs:156]: In my review I had in mind function `ensure_admin` that just checks that message sender is admin and returns Err if it's not so. This function then can be used in all methods that require admin access.

File: gear-programs/vft-treasury/src/wasm/tests/utils_gclient.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/utils_gclient.rs:12]:
```suggestion
) -> gclient::Result<(MessageId, ProgramId)> {
```

File: gear-programs/vft-treasury/src/wasm/tests/utils_gclient.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/utils_gclient.rs:36]:
```suggestion
) -> gclient::Result<ProgramId> {
```

File: gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs:8]: Pleaqse don't use macros for it

File: gear-programs/vft-treasury/src/wasm/tests/utils_gclient.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/utils_gclient.rs:8]: As I see you don't use it anywhere except `upload_program`, so I think it can be only 1 function

File: gear-programs/vft-treasury/src/wasm/tests/utils_gclient.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/utils_gclient.rs:48]: Macro seems excess here. I suggest re-write it as function or remove it completely

File: gear-programs/vft-treasury/src/wasm/tests/utils_gclient.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/utils_gclient.rs:120]: Can it also be done without macros?

File: gear-programs/vft-treasury/src/wasm/tests/end2end.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/end2end.rs:25]: Better to have `assert!` here

File: gear-programs/vft-treasury/src/wasm/tests/end2end.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/end2end.rs:52]: Assert here too?

**Review by mertwole — CHANGES_REQUESTED** [review]

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (playX18) [reply gear-programs/vft-treasury/src/services/mod.rs]: Should then update_admin also exist or its not necessary?

File: gear-programs/vft-treasury/src/wasm/tests/utils_gclient.rs
Comment (mertwole) [reply gear-programs/vft-treasury/src/wasm/tests/utils_gclient.rs]: I think it's better as macros are less readable, especially at calling side

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (mertwole) [reply gear-programs/vft-treasury/src/services/mod.rs]: I think no because it will consist of 1 line of code

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/mod.rs:156]:
```suggestion
    pub fn ensure_admin(&self) -> Result<(), Error> {
```

File: gear-programs/vft-treasury/src/services/mod.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/services/mod.rs:167]: You're still using this check everywhere but it can be replaced with `ensure_admin`

File: gear-programs/vft-treasury/src/wasm/tests/end2end.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/end2end.rs:9]:
```suggestion
    // It will not work on local node as vft-treasury logic relies on pallet-gear-eth-bridge
    // which will be initialized only in ~12 hrs from start on local node.
    let mut client = GearApi::vara_testnet().await?;
```

File: gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs:8]: It's not clear in which order programs are returned. Consider introducing new `struct` for it

File: gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs
Comment (mertwole) [inline gear-programs/vft-treasury/src/wasm/tests/vft_treasury.rs:16]: I suppose it shouldn't fai by definition

**Review by mertwole — APPROVED** [review]

**Review by ByteNacked — APPROVED** [review]

**Pull request merged**
